### PR TITLE
Handle Patreon's vimeo embeds

### PR DESCRIPTION
**Pull request by rhsmachine**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
Fix vimeo embeds. Closes https://github.com/yt-dlp/yt-dlp/issues/274

### CODE REVIEW

File: yt_dlp/extractor/patreon.py [hunk 1]
```diff
@@ -137,13 +139,24 @@ def _real_extract(self, url):
                     })
 
         if not info.get('url'):
-            embed_url = try_get(attributes, lambda x: x['embed']['url'])
-            if embed_url:
+            # handle Vimeo embeds
+            if attributes.get('embed').get('provider') == 'Vimeo':
```
Comment (pukkandan) [inline yt_dlp/extractor/patreon.py:143]: use `try_get` instead of `.get().get()`

File: yt_dlp/extractor/patreon.py [hunk 2]
```diff
@@ -137,13 +139,24 @@ def _real_extract(self, url):
...
+                embed_html = attributes.get('embed').get('html')
```
Comment (pukkandan) [inline yt_dlp/extractor/patreon.py:144]: same as above

File: yt_dlp/extractor/patreon.py [hunk 3]
```diff
@@ -137,13 +139,24 @@ def _real_extract(self, url):
...
+                v_url = self._search_regex(r'''src=(https%3A%2F%2Fplayer\.vimeo\.com.+)%3F''', embed_html, "vimeo url")
                 info.update({
-                    '_type': 'url',
-                    'url': embed_url,
+                    '_type': 'url_transparent',
+                    'url': VimeoIE._smuggle_referrer(compat_urllib_parse_unquote(v_url), 'https://patreon.com'),
+                    'ie_key': 'Vimeo',
                 })
```
Comment (pukkandan) [inline yt_dlp/extractor/patreon.py:150]:
```suggestion
                v_url = url_or_none(compat_urllib_parse_unquote(
                    self._search_regex(r'src=(https%3A%2F%2Fplayer\.vimeo\.com.+)%3F', embed_html, 'vimeo url', fatal=False)))
                if v_url:
                    info.update({
                        '_type': 'url_transparent',
                        'url': VimeoIE._smuggle_referrer(v_url, 'https://patreon.com'),
                        'ie_key': 'Vimeo',
                    })
```

**(untested)**

Basically, use `fatal=False` and check for the validity of the url

I also changed the quotes to `'` just for consistency

File: yt_dlp/extractor/patreon.py [hunk 4]
```diff
@@ -137,13 +139,24 @@ def _real_extract(self, url):
...
 
+            else:
```
Comment (pukkandan) [inline yt_dlp/extractor/patreon.py:152]:
```suggestion
            if not info.get('url'):
```

This is how it is done in the other places

File: yt_dlp/extractor/patreon.py [hunk 5]
```diff
@@ -12,6 +12,8 @@
     str_or_none,
     try_get,
 )
+from ..compat import compat_urllib_parse_unquote
+from .vimeo import VimeoIE
```
Comment (pukkandan) [inline yt_dlp/extractor/patreon.py:16]: Sort the imports. like:

```py
from .common import InfoExtractor
from .vimeo import VimeoIE

from ..compat import compat_urllib_parse_unquote
from ..utils import (
    clean_html,
    determine_ext,
    int_or_none,
    KNOWN_EXTENSIONS,
    mimetype2ext,
    parse_iso8601,
    str_or_none,
    try_get,
)
```

**Review by pukkandan — COMMENTED** [review]
Also, add a test

**Comment by TheFrenchGhosty** [comment]
Tested on an individual video, works great:

```
yt-dlp --verbose --cookies patreon.com_cookies.txt https://www.patreon.com/posts/[REDACTED]
[debug] Command-line config: ['--verbose', '--cookies', 'patreon.com_cookies.txt', 'https://www.patreon.com/posts/[REDACTED]']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] yt-dlp version 2021.05.20 
[debug] Python version 3.7.3 (CPython 64bit) - Linux-4.19.0-12-amd64-x86_64-with-debian-10.6
[debug] exe versions: ffmpeg 4.1.6-1, ffprobe 4.1.6-1
[debug] Proxy map: {}
[debug] [Patreon] Extracting URL: https://www.patreon.com/posts/[REDACTED]
[Patreon] 51706779: Downloading JSON metadata
[debug] [vimeo] Extracting URL: https://player.vimeo.com/video/[REDACTED]#__youtubedl_smuggle=%7B%22http_headers%22%3A+%7B%22Referer%22%3A+%22https%3A%2F%2Fpatreon.com%22%7D%7D
[vimeo] 555089736: Downloading webpage
[vimeo] 555089736: Extracting information
[vimeo] 555089736: Downloading JSON metadata
WARNING: [vimeo] Unable to download JSON metadata: HTTP Error 404: Not Found
[vimeo] 555089736: Downloading akfire_interconnect_quic m3u8 information
[vimeo] 555089736: Downloading akfire_interconnect_quic m3u8 information
[vimeo] 555089736: Downloading akfire_interconnect_quic MPD information
[vimeo] 555089736: Downloading akfire_interconnect_quic MPD information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 555089736: Downloading 1 format(s): hls-akfire_interconnect_quic-5324
[debug] Invoking downloader on "[REDACTED]"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 498
[download] Destination: [REDACTED].mp4
[download]   2.0% of ~1.79GiB at  1.84MiB/s ETA 17:24
```

**Comment by TheFrenchGhosty** [comment]
Tested on an individual video, works great:

```
yt-dlp --verbose --cookies patreon.com_cookies.txt https://www.patreon.com/posts/[REDACTED]
[debug] Command-line config: ['--verbose', '--cookies', 'patreon.com_cookies.txt', 'https://www.patreon.com/posts/[REDACTED]']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] yt-dlp version 2021.05.20 
[debug] Python version 3.7.3 (CPython 64bit) - Linux-4.19.0-12-amd64-x86_64-with-debian-10.6
[debug] exe versions: ffmpeg 4.1.6-1, ffprobe 4.1.6-1
[debug] Proxy map: {}
[debug] [Patreon] Extracting URL: https://www.patreon.com/posts/[REDACTED]
[Patreon] 51706779: Downloading JSON metadata
[debug] [vimeo] Extracting URL: https://player.vimeo.com/video/[REDACTED]#__youtubedl_smuggle=%7B%22http_headers%22%3A+%7B%22Referer%22%3A+%22https%3A%2F%2Fpatreon.com%22%7D%7D
[vimeo] [REDACTED]: Downloading webpage
[vimeo] [REDACTED]: Extracting information
[vimeo] [REDACTED]: Downloading JSON metadata
WARNING: [vimeo] Unable to download JSON metadata: HTTP Error 404: Not Found
[vimeo] [REDACTED]: Downloading akfire_interconnect_quic m3u8 information
[vimeo] [REDACTED]: Downloading akfire_interconnect_quic m3u8 information
[vimeo] [REDACTED]: Downloading akfire_interconnect_quic MPD information
[vimeo] [REDACTED]: Downloading akfire_interconnect_quic MPD information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] [REDACTED]: Downloading 1 format(s): hls-akfire_interconnect_quic-5324
[debug] Invoking downloader on "[REDACTED]"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 498
[download] Destination: [REDACTED].mp4
[download]   2.0% of ~1.79GiB at  1.84MiB/s ETA 17:24
```

**Comment by TheFrenchGhosty** [comment]
Tested on a profile, goes back to the generic extractor:

```
yt-dlp --verbose --cookies patreon.com_cookies.txt https://www.patreon.com/[REDACTED]/
[debug] Command-line config: ['--verbose', '--cookies', 'patreon.com_cookies.txt', 'https://www.patreon.com/[REDACTED]/']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] yt-dlp version 2021.05.20 
[debug] Python version 3.7.3 (CPython 64bit) - Linux-4.19.0-12-amd64-x86_64-with-debian-10.6
[debug] exe versions: ffmpeg 4.1.6-1, ffprobe 4.1.6-1
[debug] Proxy map: {}
[debug] [generic] Extracting URL: https://www.patreon.com/[REDACTED]/
[generic] [REDACTED]: Requesting header
[redirect] Following redirect to https://www.patreon.com/[REDACTED]/posts
[debug] [generic] Extracting URL: https://www.patreon.com/[REDACTED]/posts
[generic] posts: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] posts: Downloading webpage
[generic] posts: Extracting information
ERROR: Unsupported URL: https://www.patreon.com/[REDACTED]/posts
Traceback (most recent call last):
  File "/home/vm/.local/lib/python3.7/site-packages/yt_dlp/YoutubeDL.py", line 1128, in wrapper
    return func(self, *args, **kwargs)
  File "/home/vm/.local/lib/python3.7/site-packages/yt_dlp/YoutubeDL.py", line 1149, in __extract_info
    ie_result = ie.extract(url)
  File "/home/vm/.local/lib/python3.7/site-packages/yt_dlp/extractor/common.py", line 564, in extract
    ie_result = self._real_extract(url)
  File "/home/vm/.local/lib/python3.7/site-packages/yt_dlp/extractor/generic.py", line 3573, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.patreon.com/[REDACTED]/posts
```

File: yt_dlp/extractor/patreon.py [hunk 6]
```diff
@@ -63,6 +67,20 @@ class PatreonIE(InfoExtractor):
     }, {
         'url': 'https://www.patreon.com/posts/743933',
         'only_matching': True,
+    }, {
+        'url': 'https://www.patreon.com/posts/kitchen-as-seen-51706779',
+        'md5': '96656690071f6d64895866008484251b',
+        'info_dict': {
+            'id': '555089736',
+            "ext": "mp4",
```
Comment (pukkandan) [inline yt_dlp/extractor/patreon.py:75]:
```suggestion
            'ext': 'mp4',
```

**Review by pukkandan — APPROVED** [review]

**Comment by pukkandan** [comment]
> Tested on a profile, goes back to the generic extractor (it's maybe out of scope of this PR?):


Yes, this PR is only for single videos

**Pull request merged**